[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Jibby2k1/SPS_Curriculum/blob/main/Intro_Math/Analysis/Random_Variables.ipynb)


Content for and by IEEE Signal Processing Society. (Raul Valle & Contributors)

# Random Variables

> ⚠️ **Draft — pending instructor review.** The visuals below execute, but execution cannot verify proofs. An SPS instructor should vet the arguments in this notebook before it is taught. Remove this banner after review.

Probability is not a new subject — it is [Measure Theory](./Measure_Theory.ipynb) with total mass 1. This workshop makes that identification precise: random variables are measurable functions, expectation is the Lebesgue integral, and every distribution you've ever met is a pushforward measure. The reward: one clean framework under the noise models of [Kalman filtering](../../Intro_Time_Series/Intro_AdFilt_KF.ipynb) and the losses of [machine learning](../../Intro_Mach_Learn/README.md).

### Visual setup & helpers
*(Safe to re-run anytime.)*

In [1]:
import numpy as np
import matplotlib.pyplot as plt
rng = np.random.default_rng(0)

---
### 🕐 Session 1 of 2 — *Probability Spaces & Random Variables* (~35 min)
**Goal:** define probability as a measure; define random variables as measurable functions and distributions as pushforwards.
**Builds on:** [Measure Theory](./Measure_Theory.ipynb). &nbsp; **Feeds into:** Session 2 (distributions, expectation, moments).

---

<details>
<summary>🎓 <b>Teacher notes — Session 1: Probability Spaces & Random Variables</b></summary>

**Timing (~35 min).** 8 min probability as a measure · 10 min the "not random, not a variable" reframe · 8 min pushforward laws · 9 min the CDF theorem and demo.

**Board first — state the identification and mean it.** Probability is [Measure Theory](./Measure_Theory.ipynb) with total mass 1. Nothing else is new. Every theorem from that workshop transfers immediately — monotonicity, countable additivity, continuity from below and above — and the last one comes *free of the finiteness caveat*, because everything is bounded by 1. Students who see probability as a separate subject carry twice the load; make the transfer explicit in the first two minutes.

**The reframe that fixes the most confusion.** A random variable is **not random and not a variable**: it is a fixed, deterministic function $X : \Omega \to \mathbb{R}$. All the randomness lives in which $\omega$ the world serves up. Say it, write it, and return to it — students who internalise this stop being puzzled by conditional expectation, by $\sigma(X)$, and by martingales later in the track.

**Then "measurable" as fine print with a purpose.** It is not a technical nuisance: it is the requirement that $\{X \le x\}$ be an *askable question*, so that its probability exists at all. If the preimage is not in $\mathcal{F}$, $P(X \le x)$ is literally undefined. Point at Vitali's set from Measure Theory for why $\mathcal{F}$ cannot simply be all subsets.

**The pushforward is the licence to forget $\Omega$, and that is worth saying out loud.** $P_X(B) = P(X^{-1}(B))$ moves everything to the real line, which is why nobody in practice ever names their sample space. Ask the room what $\Omega$ is for a coin flip, for a Gaussian, for tomorrow's temperature — it does not matter, and the pushforward is the theorem that says so.

**Do the CDF proof; all three parts are one line each.** Monotone from monotonicity of $P$; right-continuous from continuity from above; the limits from continuity again. Then the important remark: **left-continuity can fail**, and the jumps are exactly the atoms, with jump size $P(X = x)$. That is what the demo shows.

**Use the three-panel demo to break a false dichotomy.** Students arrive believing distributions are either discrete or continuous. The third panel is neither — a mixture — and it is perfectly legitimate. Ask for a real example: rainfall (an atom at zero, a density above it), insurance claims, time-to-event data with censoring. The measure-theoretic definition covers all three without special cases, which is precisely why it is worth the abstraction.
</details>

## Probability Spaces

**Definition.** A *probability space* is a measure space $(\Omega, \mathcal{F}, P)$ with $P(\Omega) = 1$:

- $\Omega$ — the *sample space*: every way the experiment could turn out;
- $\mathcal{F}$ — a σ-algebra of *events* (the askable questions);
- $P$ — a measure assigning each event its probability.

Everything from [Measure Theory](./Measure_Theory.ipynb) transfers instantly: monotonicity, countable (sub)additivity, continuity from below/above (no finiteness caveat needed — everything is $\le 1$).

💡 **Intuition.** Why must $\mathcal{F}$ be a σ-algebra and not "all subsets"? Two reasons. Philosophically, $\mathcal{F}$ encodes *what is knowable* — you can ask about complements ("not A") and countable unions ("eventually some $A_n$"), which is exactly closure under σ-operations. Technically, for continuous experiments (Ω = [0,1]) Vitali's set already showed that measuring everything is impossible.

## Random Variables

💡 **Intuition.** A random variable is **not random and not a variable** — it's a fixed, deterministic *function* $X: \Omega \to \mathbb{R}$ that reports a number about the outcome. All the randomness lives in which $\omega \in \Omega$ the world serves up. "Measurable" is the fine print making the function compatible with the askable questions: events like $\{X \le x\}$ must belong to $\mathcal{F}$, or their probability is undefined.

**Definition.** $X: \Omega \to \mathbb{R}$ is a *random variable* if it is $\mathcal{F}$-measurable: $X^{-1}(B) \in \mathcal{F}$ for every Borel set $B \in \mathcal{B}(\mathbb{R})$. (Checking $\{X \le x\} \in \mathcal{F}$ for all $x$ suffices, since the half-lines generate $\mathcal{B}$.)

**Definition (distribution).** The *law* of $X$ is the pushforward measure on $\mathbb{R}$:

$$P_X(B) = P(X^{-1}(B)) = P(X \in B).$$

$P_X$ is a probability measure on $(\mathbb{R}, \mathcal{B})$ — check the axioms via preimages. This is the licence to forget $\Omega$: all computations about $X$ alone happen on the real line with $P_X$.

### Proof: The CDF characterizes the law

Define $F(x) = P(X \le x)$. Then $F$ is (1) non-decreasing, (2) right-continuous, and (3) has limits $0$ at $-\infty$ and $1$ at $+\infty$.

(1) $x \le y \implies \{X \le x\} \subseteq \{X \le y\}$; apply monotonicity.

(2) Let $x_n \downarrow x$. The events $\{X \le x_n\}$ decrease to $\bigcap_n \{X \le x_n\} = \{X \le x\}$; continuity from above gives $F(x_n) \to F(x)$.

(3) $\{X \le -n\} \downarrow \emptyset$ and $\{X \le n\} \uparrow \Omega$; apply continuity again. $\blacksquare$

*(Conversely, any $F$ with these properties is the CDF of some random variable — and since intervals generate $\mathcal{B}$, the CDF pins down the whole law. Note left-continuity can fail: $F$ jumps exactly at atoms, by size $P(X = x)$.)*

In [2]:
# One function, three personalities: CDFs of discrete, continuous, and mixed laws
x = np.linspace(-0.5, 3.5, 2000)

F_disc = np.select([x < 0, x < 1, x < 2], [0, 0.3, 0.8], default=1.0)       # atoms at 0,1,2
F_cont = np.clip(x / 3, 0, 1)                                                # Uniform[0,3]
F_mix  = 0.5 * F_disc + 0.5 * F_cont                                         # half-and-half

fig, axes = plt.subplots(1, 3, figsize=(10, 2.6), sharey=True)
for ax, F, t in zip(axes, [F_disc, F_cont, F_mix], ["discrete (jumps only)", "continuous", "mixed"]):
    ax.plot(x, F); ax.set_title(t); ax.grid(True)
axes[0].set_ylabel("F(x)")
plt.suptitle("Right-continuous, non-decreasing, 0→1 — that's ALL a CDF must be", y=1.05)
plt.tight_layout(); plt.show()

/tmp/ipykernel_1898601/641952581.py:13: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.tight_layout(); plt.show()


**What just happened.** Three CDFs, three completely different kinds of distribution — and all three satisfy the *same* three properties: non-decreasing, right-continuous, running from 0 to 1.

**The third panel is the one that matters.** Students arrive believing distributions are either discrete or continuous, and the mixed case looks like a pathology. It is not: it is ordinary. Rainfall has an atom at zero (many dry days) and a density above it. Insurance claims have an atom at zero and a continuous tail. Time-to-event data with censoring is mixed by construction. The measure-theoretic framework covers all three with **no special cases**, which is the entire payoff for the abstraction.

**Read the jumps precisely.** $F$ is right-continuous everywhere and **left-continuity fails exactly at atoms**, where the jump size is $P(X = x)$. In the left panel the jumps at 0, 1, 2 have heights 0.3, 0.5, 0.2 — those *are* the point probabilities. So the CDF is not merely a summary; it encodes the atoms explicitly in its discontinuities, which a density cannot do.

**And that is why the CDF characterises the law.** Half-lines $(-\infty, x]$ generate the Borel σ-algebra, so knowing $P(X \le x)$ for every $x$ determines $P_X(B)$ for *every* Borel set. One function of one real variable pins down the entire distribution. The three properties proved above are not just necessary — they are sufficient, so any function with them is the CDF of some random variable.

**Why this is more general than a density.** A density requires the law to be absolutely continuous with respect to Lebesgue measure, and the left and right panels are not. Worse, there are laws with *no* atoms and *no* density — the Cantor distribution is the standard example, continuous but supported on a measure-zero set. The CDF handles all of these; $f$ does not. That is why the measure-theoretic definition is primary and the density is a special case.

**Keep the framing from the intuition cell.** A random variable is a deterministic function $X : \Omega \to \mathbb{R}$; the randomness is in which $\omega$ occurs. The **pushforward** $P_X(B) = P(X^{-1}(B))$ then moves everything onto the real line — which is the formal licence for the fact that nobody in practice ever names their sample space. Ask what $\Omega$ is for a Gaussian; it does not matter, and the pushforward is the theorem that says it does not.

---
### 🕐 Session 2 of 2 — *Distributions, Expectation & Moments* (~40 min)
**Goal:** compute with densities; define expectation as a Lebesgue integral; prove Markov & Chebyshev.
**Builds on:** Session 1. &nbsp; **Feeds into:** [Independence](./README.md#6-independence-draft--pending-review).

---

<details>
<summary>🎓 <b>Teacher notes — Session 2: Distributions, Expectation & Moments</b></summary>

**Timing (~40 min).** 8 min densities and their limits · 10 min expectation as the Lebesgue integral · 10 min LOTUS and the proof ladder · 12 min Markov, Chebyshev, and the demos.

**Board first — expectation is *literally* the Lebesgue integral.** $E[X] = \int_\Omega X\,dP$. Not analogous to it, not a probabilistic version of it — the same object from [Measure Theory](./Measure_Theory.ipynb), applied to a measure of total mass 1. For a coin-flip payout it collapses to the weighted sum students learned as children; the machinery exists so that *one* definition covers payouts with densities, with atoms, or with neither.

**Name the practical payoff of that generality.** Because expectation is a Lebesgue integral, MCT and DCT apply — so you can swap limits and expectations under stated conditions. Every "$\lim E = E \lim$" step in later probability arguments is one of those theorems, and none of them would be available if expectation had been defined separately for discrete and continuous cases.

**Teach the proof ladder explicitly, because it recurs.** LOTUS is proved by: indicators (true by definition of pushforward) → simple functions (linearity) → non-negative (MCT) → general (split into $g^\pm$). Tell the room this **indicator → simple → limit** pattern is *the* standard proof shape in measure theory and that they will see it constantly. Recognising the ladder makes a dozen later proofs feel routine.

**Markov's proof is one line and worth doing at the board.** Pointwise, $a\mathbf{1}_{\{X\ge a\}} \le X$ — check both cases aloud — then take expectations and use monotonicity of the integral. Chebyshev is then Markov applied to $(X - EX)^2$. Two of the most-used inequalities in the subject, from one pointwise comparison.

**Frame the Chebyshev demo as "loose, and that is the point."** The bound is far above the Gaussian tail — at $a = 3$ it gives 0.111 against a true 0.0027, a factor of 41. Ask why anyone would use it. Because it needs **only a variance**: no distributional assumption at all. Chebyshev is a guarantee that holds for every distribution with that variance, including the worst one, and being valid for the adversarial case is exactly what makes it loose for a Gaussian. Same argument as [Concentration](../Concentration/Concentration_Inequalities.ipynb) Session 1, and this is where that workshop starts.

**Markov's 6.7× looseness in the final cell makes the same point more cheaply.** It uses only the *mean* — even less information than Chebyshev — so it is looser still. Have the room rank the three (Markov, Chebyshev, sub-Gaussian) by how much they assume and watch the tightness follow the same order. **More assumptions, tighter bound** is the entire economy of concentration inequalities.

**On the sampling cell.** Sample mean 1.004 and variance 1.025 against theory 1.000 — good agreement, and worth flagging that this *illustrates* the law of large numbers rather than proving it. The proof is the next workshop, and the honest framing is that we are checking the machinery behaves before we justify why it must.
</details>

## Densities

**Definition.** $X$ has *density* $f$ if $P(X \in B) = \int_B f \, d\lambda$ for all Borel $B$ — the law is "Lebesgue measure reweighted by $f$." Then $F' = f$ wherever $f$ is continuous. Not every law has one (atoms don't; stranger things — laws supported on the Cantor set — don't either), which is why the measure-theoretic definition, not the density, is primary.

## Expectation

💡 **Intuition.** Expectation is *literally* the Lebesgue integral of the previous notebook: $E[X] = \int_\Omega X \, dP$ — the y-axis-slicing integral, built from simple functions up. For a coin-flip payout it reduces to the weighted sum you learned as a kid; the machinery exists so that the *same* definition covers payouts with densities, atoms, or neither — and so MCT/DCT let us swap limits and expectations.

**Change of variables (LOTUS).** For measurable $g$ with $g(X) \in L^1$:

$$E[g(X)] = \int_\Omega g(X(\omega)) \, dP(\omega) = \int_{\mathbb{R}} g(x) \, dP_X(x) \;\; \Big( = \int g(x) f(x) \, dx \text{ if a density exists} \Big)$$

*Proof route:* true for $g = \mathbf{1}_B$ by definition of pushforward; extend to simple $g$ by linearity, to $g \ge 0$ by MCT, to general $g$ by splitting $g^{\pm}$. This "indicator → simple → limit" ladder is *the* standard proof pattern of measure theory — you'll use it constantly.

**Moments.** $E[X^k]$ is the $k$-th moment; variance is $\mathrm{Var}(X) = E[(X - E[X])^2] = E[X^2] - (E[X])^2$ (expand the square, use linearity).

### Proof: Markov & Chebyshev inequalities

**Markov.** For $X \ge 0$ and $a > 0$: $\;P(X \ge a) \le \frac{E[X]}{a}$.

Pointwise, $a \, \mathbf{1}_{\{X \ge a\}} \le X$ (if $X(\omega) \ge a$ the left side is $a$; otherwise it's $0$). Take expectations — monotonicity of the integral gives $a \, P(X \ge a) \le E[X]$. $\blacksquare$

**Chebyshev.** Apply Markov to the non-negative variable $(X - E[X])^2$ with threshold $a^2$:

$$P(|X - E[X]| \ge a) = P\big((X - E[X])^2 \ge a^2\big) \le \frac{\mathrm{Var}(X)}{a^2}. \;\blacksquare$$

One line each — and Chebyshev is the engine of the weak law of large numbers in [Independence](./README.md#6-independence-draft--pending-review).

In [3]:
# How tight is Chebyshev? Empirical tail vs the bound, for a standard normal
a_grid = np.linspace(0.5, 4, 30)
samples = rng.standard_normal(200_000)
empirical = [(np.abs(samples) >= a).mean() for a in a_grid]

plt.figure(figsize=(7.5, 3))
plt.semilogy(a_grid, empirical, label="empirical  P(|X| ≥ a),  X ~ N(0,1)")
plt.semilogy(a_grid, np.minimum(1, 1 / a_grid**2), "--", label="Chebyshev bound  1/a²")
plt.legend(); plt.grid(True); plt.xlabel("a")
plt.title("Chebyshev is loose for Gaussians — its power is needing ONLY a variance")
plt.tight_layout(); plt.show()

/tmp/ipykernel_1898601/1205581157.py:11: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.tight_layout(); plt.show()


**What just happened.** The Chebyshev bound sits far above the true Gaussian tail — and the gap widens as $a$ grows. At $a = 2$ the bound is 0.25 against a true 0.046; at $a = 3$ it is 0.111 against 0.0027, a factor of **41**; at $a = 4$ the gap is over 300×.

**The looseness is structural, not a defect.** Chebyshev uses **only the variance**. It carries no information about shape, tails, symmetry, or anything else — so it must be valid for *every* distribution with that variance, including the worst one. And the worst one is not a Gaussian: it is a two-point distribution placing all its mass at $\pm a$, for which Chebyshev is **tight**. A bound that survives the adversarial case is necessarily generous for a well-behaved one.

**Which is exactly why it is worth having.** Ask what you need to know to use it: the variance, and nothing else. No normality, no unimodality, no bounded support. That is an extraordinarily cheap hypothesis, and it is why Chebyshev appears in proofs — including the weak law of large numbers in [Independence](./README.md#6-independence-draft--pending-review), where the entire argument is "variance shrinks like $1/n$, therefore the average concentrates."

**Note the shape of the decay, because it is what gets improved later.** Chebyshev falls like $1/a^2$ — *polynomially*. The Gaussian tail falls like $e^{-a^2/2}$ — *exponentially*. On the log axis that is a slowly-flattening line against one that keeps bending downward, and no constant factor closes a gap between different functional forms.

Improving it requires assuming more. [Concentration Inequalities](../Concentration/Concentration_Inequalities.ipynb) does exactly that: assume boundedness, and Hoeffding's lemma buys sub-Gaussian tails via the Chernoff trick — Markov applied to $e^{\lambda X}$ rather than to $X$. **More assumptions, tighter bound**, and that trade is the whole economy of the subject.

**And both inequalities come from one pointwise observation.** Markov is $a\mathbf{1}_{\{X \ge a\}} \le X$, true by checking two cases, then integrated. Chebyshev is Markov applied to the non-negative variable $(X - E[X])^2$ at threshold $a^2$. Two of the most-used results in probability, from a single comparison of functions and the monotonicity of the Lebesgue integral.

### Sampling meets the theory

A density is not a histogram — but histograms of samples converge to the density, and empirical means converge to $E[X]$ (that's the law of large numbers, proven properly next workshop).

In [4]:
# Exponential(1): compare sample histogram/mean/variance to the theory
X = rng.exponential(1.0, size=100_000)
x = np.linspace(0, 6, 300)

plt.figure(figsize=(7.5, 2.8))
plt.hist(X, bins=80, range=(0, 6), density=True, alpha=0.5, label="100k samples")
plt.plot(x, np.exp(-x), "k", label="density $e^{-x}$")
plt.legend(); plt.title("Exponential(1)")
plt.tight_layout(); plt.show()

print(f"E[X]:   theory 1.000   sample {X.mean():.3f}")
print(f"Var(X): theory 1.000   sample {X.var():.3f}")
print(f"P(X≥3): theory {np.exp(-3):.4f}  sample {(X >= 3).mean():.4f}  Markov bound {1/3:.3f}")

E[X]:   theory 1.000   sample 1.004
Var(X): theory 1.000   sample 1.025
P(X≥3): theory 0.0498  sample 0.0511  Markov bound 0.333


/tmp/ipykernel_1898601/2728721527.py:9: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.tight_layout(); plt.show()


**What just happened.** Theory and samples agree across the board:

| quantity | theory | 100k samples |
|---|---|---|
| $E[X]$ | 1.000 | 1.004 |
| $\mathrm{Var}(X)$ | 1.000 | 1.025 |
| $P(X \ge 3)$ | 0.0498 | 0.0511 |

And the histogram sits on $e^{-x}$ — the abstract machinery of pushforward measures and Lebesgue integration produces numbers that a random number generator reproduces.

**Now compare the last row against its bound.** Markov gives $P(X \ge 3) \le E[X]/3 = 0.333$, against a true 0.0498 — **6.7× loose**. Put that beside Chebyshev's 41× looseness at the same threshold and rank the three tools by how much they assume:

| bound | uses | looseness at $a = 3$ |
|---|---|---|
| Markov | the mean only | 6.7× (here) |
| Chebyshev | mean + variance | 41× (Gaussian) |
| sub-Gaussian | boundedness / MGF | far tighter |

**More assumptions, tighter bound.** That ordering is the entire economy of concentration inequalities, and [Concentration Inequalities](../Concentration/Concentration_Inequalities.ipynb) is the workshop that walks up it. Markov is the cheapest possible statement — one number about the distribution — and it is correspondingly weak; every improvement in that track is purchased with an extra hypothesis.

Worth noting that Markov's apparent tightness here is partly luck of the threshold: it is loose by 6.7× at $a = 3$ and would be far looser at $a = 6$, because it decays only like $1/a$ while the exponential tail decays like $e^{-a}$. Again, different functional forms, and no constant closes that.

**Be clear about what this cell proves and what it merely illustrates.** The agreement between sample statistics and theory is *not* a verification of the sampler or the mathematics; it is a demonstration that the two are consistent. **Why** the sample mean should approach $E[X]$ at all is the law of large numbers, which is proved in the [next workshop](./README.md#6-independence-draft--pending-review). Right now we are checking that the machinery behaves before justifying why it must — an honest ordering, and worth saying so rather than implying the convergence has been established.

Note also which theorem is being used silently: computing $E[X] = \int x e^{-x}dx$ from the *density* rather than from $\int_\Omega X\,dP$ over an unnamed sample space is **LOTUS**, the change-of-variables result proved above by the indicator → simple → limit ladder. Every practical expectation calculation you will ever do is that theorem, applied without comment.

---
## Where next

- [Independence](./README.md#6-independence-draft--pending-review) — product measures, Borel–Cantelli, and the laws of large numbers that justify every Monte Carlo estimate above.
- [Adaptive Filtering: Kalman](../../Intro_Time_Series/Intro_AdFilt_KF.ipynb) — the $Q$, $R$ matrices are covariances of exactly these objects.
- [Measure Theory](./Measure_Theory.ipynb) — the machinery, if you skipped ahead.